lib/fdtable.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief File descriptor table
+ *
+ * This file provides generic file descriptor table implementation, suitable
+ * for any I/O object implementing POSIX I/O semantics (i.e. read/write +
+ * aux operations).
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <misc/fdtable.h>
+
+struct fd_entry {
+	void *obj;
+	const struct fd_op_vtable *vtable;
+};
+
+#ifdef CONFIG_POSIX_API
+static struct fd_op_vtable stdinout_fd_op_vtable;
+#endif
+
+static struct fd_entry fdtable[CONFIG_POSIX_MAX_FDS] =
+{
+#ifdef CONFIG_POSIX_API
+	/*
+	 * Predefine entries for stdin/stdout/stderr. Object pointer
+	 * is unused and just should be !0.
+	 */
+	{(void*)0x10, &stdinout_fd_op_vtable},
+	{(void*)0x11, &stdinout_fd_op_vtable},
+	{(void*)0x12, &stdinout_fd_op_vtable},
+#endif
+};
+
+static int _find_fd_entry(void)
+{
+	int fd;
+
+	for (fd = 0; fd < ARRAY_SIZE(fdtable); fd++) {
```

Comment:
suggestion: at a cost of a little increased memory, we could define a bitfield to look for unused descriptor IDs, which would be faster than iterating over the array.

Reply:
A thing to keep in mind here: if `dup2()` ends up being implemented, one can say "duplicate fd#1 into fd#9000".  So using `fd` as an index in an array, or using the bitfield to determine which FDs are being used, aren't going to work.

Given that the number of descriptors in a Zephyr thread will be small, a linear search through the table is good enough; sure, the bitfield can be used there as an optimization to speed up looking for a new fd when allocating, or to clean up when a thread ends, but it might be a better idea to just keep the table sorted and do a binary search.  (Keeping it ordered has an added cost of poking a hole in the array once an item is added, or moving everything back together when an item is removed, but again, N is small here, and file descriptor->file description lookup is hotter than creating/destroying a file descriptor.)

Reply:
Yeah, definitely let's leave optimizations for later. I for example thought of having inplace freelist, but of course, that doesn't follow POSIX fd allocation semantics (looked up that it's mandated for open(), not just for ol' good dup()).

Reply:
> A thing to keep in mind here: if dup2() ends up being implemented, one can say "duplicate fd#1 into fd#9000"

I wouldn't think it works like that, you could specify any fd no., but within bounds of max supported number of fd's. And at this stage, I'd prefer to remain at the happy state of thinking that POSIX is less a bloat that it might be ;-).